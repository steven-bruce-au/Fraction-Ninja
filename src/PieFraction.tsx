import React from 'react';
import { PieChart, Pie, Cell, ResponsiveContainer } from 'recharts';
import styles from './PieFraction.module.scss';

/** Data point for use by recharts */
class DataPoint {
  name: string
  value: number

  constructor(name: string, value: number) {
    this.name = name;
    this.value = 1;
  }
}

type PieState = {
  data: Array<DataPoint>
}

type PieProps = {
  numerator: number,
  denominator: number,
  equal: boolean
}

/** Renders a pie chart using the D3 library via recharts
*/
export default class PieFraction extends React.Component<PieProps, PieState> {

  constructor(props: any) {
    super(props);
    this.state = {
      data: this.fractionData(props.numerator,props.denominator)
    }
  }

  /** Handle an updated chart value
  @param prevProps the previous properties
  */
  componentDidUpdate(prevProps: PieProps) {
      if(prevProps.numerator !== this.props.numerator || prevProps.denominator !== this.props.denominator) {
        this.setState({ data: this.fractionData(this.props.numerator, this.props.denominator)});
      }
  }

  /** construct an array of data points
  @param numerator The value of the numerator
  @param denominator The value of the denominator
  @return an array of data points
  */
  fractionData(numerator: number, denominator: number) : Array<DataPoint> {
    if(denominator > 0) {
      const array = this.fractionData(numerator, denominator - 1);
      array.push(new DataPoint(numerator.toString(), denominator));
      return array;
    } else {
      return new Array<DataPoint>();
    }
  }

  /** Determine the colour of the pie slice
  @param index The pie slice
  */
  colour(index:number): string {
    if(this.props.equal) {
      // green or grey
      return index < this.props.numerator ? "#0e0" : "#9a9"
    } else {
      // red or grey
      return index < this.props.numerator ? "#e00" : "#a99"
    }
  }

  render() {
    return (
     <div className={styles.chart}>
     <ResponsiveContainer width="100%" height="100%">
       <PieChart>
          <Pie
            data={this.state.data}
            labelLine={false}
            startAngle={90}
            endAngle={450}
            animationDuration={0}
            outerRadius={'100%'}
            fill="#8884d8"
            dataKey="value"
          >
            {
              this.state.data.map((entry, index) => <Cell key={`cell-${index}`} fill={this.colour(index)} />).reverse()
            }
          </Pie>
        </PieChart>
       </ResponsiveContainer>
      </div>
    );
  }
}
