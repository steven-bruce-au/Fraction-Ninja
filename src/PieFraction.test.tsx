import React from 'react';
import * as renderer from 'react-test-renderer';
import PieFraction from './PieFraction';
import util from 'util';


class EmptyParent implements renderer.TestRendererOptions {

   // override default implementation
   createNodeMock(element: ReactElement): any {
    const doc = document.implementation.createHTMLDocument();
    return { parentElement: doc.body };
   }
}

it('render 2/3 pie chart in red', () => {

  const component = renderer.create(<PieFraction numerator='2' denominator='3' equal='false'/> , new EmptyParent());

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();


});

