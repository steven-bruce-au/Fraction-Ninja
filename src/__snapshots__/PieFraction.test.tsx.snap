// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 2/3 pie chart in red 1`] = `
<div
  className="chart"
>
  <div
    className="recharts-responsive-container"
    style={
      Object {
        "height": "100%",
        "maxHeight": undefined,
        "minHeight": undefined,
        "minWidth": undefined,
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "display": "none",
          "height": 0,
          "position": "absolute",
          "visibility": "hidden",
          "width": 0,
        }
      }
    />
  </div>
</div>
`;
